pisa/stages/abs/pi_earth_absorption.py

```diff
@@ -0,0 +1,244 @@
+        for container in self.data:
+            container['xsection'] = self.calculate_xsections(container['flav'],
+                                                             container['nubar'],
+                                                             container['true_energy'].get(WHERE)
+                                                            )
+            calculate_survivalprob(container['rho_int'].get(WHERE),
+                                   container['xsection'].get(WHERE),
+                                   out=container['survival_prob'].get(WHERE)
+                                  )
+            container['survival_prob'].mark_changed(WHERE)
+        self.data.unlink_containers()
+      
+    @profile
+    def apply_function(self):
+        for container in self.data:
+            vectorizer.multiply(container['survival_prob'], out=container['weights'])
+            
+    def calculate_xsections(self, flav, nubar, energy):
+        '''Calculates the cross-sections on isoscalar targets.
+        The result is returned in cm^2. The xsection on one 
+        target is calculated by taking the xsection for O16
+        and dividing it by 16. 
+        '''
+        if flav == 0: pdg = 12
+        if flav == 1: pdg = 14
+        if flav == 2: pdg = 16
+        in_flavor = pdg*nubar
+        flavor = {12: 'e', -12: 'e_bar', 14: 'mu', -14: 'mu_bar', 16: 'tau', -16: 'tau_bar'}[in_flavor]
+        return (self.xsroot.Get('nu_'+flavor+'_O16').Get('tot_cc').Eval(energy)+
+                self.xsroot.Get('nu_'+flavor+'_O16').Get('tot_nc').Eval(energy))*10**(-38)/16. # this gives cm^2
+
+signatures = [
+    '(f4[::1], f4[::1], f4[::1])',
+    '(f8[::1], f8[::1], f8[::1])'
+]
+
+# TODO: make this work with the 'cuda' target. Right now, it seems like np.dot
+# does not work or is used incorrectly.
+@guvectorize(signatures, '(n),(n)->()', target=TARGET)
+def calculate_integrated_rho(layer_dists, layer_densities, out):
+    """Calculate density integrated over the path through all layers.
+    Gives the length of a matter-equivalent water column in cm.
+    
+    Parameters
+    ----------
+    layer_dists : vector
+        distance travelled through each layer
+    layer_densities : vector
+        densities of the layers
+    out : scalar
+        Result is stored here
+
+    """
+    out[0] = np.dot(layer_dists, layer_densities)*1e5 #distances are converted from km to cm
+
+@guvectorize(signatures, '(),()->()', target=TARGET)
+def calculate_survivalprob(int_rho, xsection, out):
+    """Calculate survival probability given layer distances,
+    layer densities and (pre-computed) cross-sections.
+    
+    Parameters
+    ----------
+    int_rho : scalar
+        depth of mass equivalent water column in cm
+    xsection : scalar
+        cross-section per nucleon in cm^2
+    out : scalar
+        Result is stored here
+
+    """
+    Na = 6.022E23 # nuclei per cm^(-3)
```

Comment:
I believe constants like this are avilable in pint, need to check (but that's a detail)

Reply:
Hm I didn't find it...